source/vibe/http/websockets.d

```diff
@@ -697,108 +727,234 @@ final class IncomingWebSocketMessage : InputStream {
 	}
 }
 
+/// Magic string defined by the RFC for challenging the server during upgrade
+private static immutable s_webSocketGuid = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
 
-private immutable s_webSocketGuid = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
 
 /**
- * The Opcode is 4 bytes, as defined in Section 5.2
+ * The Opcode is 4 bits, as defined in Section 5.2
  *
  * Values are defined in section 11.8
  * Currently only 6 values are defined, however the opcode is defined as
- * taking 4 bytes.
+ * taking 4 bits.
  */
-enum FrameOpcode : uint {
+enum FrameOpcode : ubyte {
 	continuation = 0x0,
 	text = 0x1,
 	binary = 0x2,
 	close = 0x8,
 	ping = 0x9,
 	pong = 0xA
 }
+static assert(FrameOpcode.max < 0b1111, "FrameOpcode is only 4 bits");
 
 
 struct Frame {
-	bool fin;
-	FrameOpcode opcode;
+	/**
+	   Contains the first byte of the frame:
+	   - The FIN bit
+	   - The 3 reserved bits RSV1, RSV2, RSV3
+	   - The 4-bits opcode
+	*/
+	ubyte fin_rsv_opcode;
```

Comment:
This could probably benefit from using `bitfields` instead of manual bit fiddling, but both are generally fine for me in this case.